sparse_strips/vello_hybrid/src/render/wgpu.rs

```diff
@@ -1020,6 +1165,173 @@ impl Programs {
+        debug_assert!(
+            self.resources.gradient_texture.width() == max_texture_dimension_2d,
+            "Gradient texture width must match max texture dimensions"
+        );
+        let current_gradient_height = self.resources.gradient_texture.height();
+        if required_gradient_height > current_gradient_height {
+            assert!(
+                required_gradient_height <= max_texture_dimension_2d,
+                "Gradient texture height exceeds max texture dimensions"
+            );
+            let required_gradient_size = (max_texture_dimension_2d * required_gradient_height) << 2; // 4 bytes per RGBA8 pixel
+            self.gradient_data
+                .resize(required_gradient_size as usize, 0);
+            let gradient_texture = Self::make_gradient_texture(
+                device,
+                max_texture_dimension_2d,
+                required_gradient_height,
+            );
+            self.resources.gradient_texture = gradient_texture;
+
+            // Since the gradient texture has changed, we need to update the gradient bind group.
+            self.resources.gradient_bind_group = Self::make_gradient_bind_group(
+                device,
+                &self.gradient_bind_group_layout,
+                &self
+                    .resources
+                    .gradient_texture
+                    .create_view(&Default::default()),
+            );
+        }
+    }
+
+    /// Update config buffer if dimensions changed.
+    fn maybe_update_config_buffer(
+        &mut self,
+        queue: &Queue,
+        max_texture_dimension_2d: u32,
+        new_render_size: &RenderSize,
+    ) {
+        if self.render_size != *new_render_size {
+            let config = Config {
+                width: new_render_size.width,
+                height: new_render_size.height,
+                strip_height: Tile::HEIGHT.into(),
+                alphas_tex_width_bits: max_texture_dimension_2d.trailing_zeros(),
+            };
+            let mut buffer = queue
+                .write_buffer_with(&self.resources.view_config_buffer, 0, SIZE_OF_CONFIG)
+                .expect("Buffer only ever holds `Config`");
+            buffer.copy_from_slice(bytemuck::bytes_of(&config));
+
+            self.render_size = new_render_size.clone();
+        }
+    }
+
+    /// Upload alpha data to the texture.
+    fn upload_alpha_texture(&mut self, queue: &Queue, alphas: &[u8]) {
+        let texture_width = self.resources.alphas_texture.width();
+        let texture_height = self.resources.alphas_texture.height();
+        debug_assert!(
+            alphas.len() <= (texture_width * texture_height * 16) as usize,
+            "Alpha texture dimensions are too small to fit the alpha data"
+        );
+
+        // After this copy to `self.alpha_data`, there may be stale trailing alpha values. These
+        // are not sampled, so can be left as-is.
+        self.alpha_data[0..alphas.len()].copy_from_slice(alphas);
```

Comment:
Happy to leave this one as a TODO since it's not directly attributable to this PR, but we also don't necessarily need to copy all the alphas into a staging buffer. We can use `alphas` directly (similar to https://github.com/linebender/vello/pull/1196/files#r2335625698) by taking the alphas from the strip generator as a vec, resizing it appropriately, truncating, and restoring it back to the generator.